tests/jvm/src/test/scala/org/typelevel/idna4s/tests/uts46/CodePointMapperPlatformTests.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2022 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package org.typelevel.idna4s.tests.uts46
+
+import cats.syntax.all._
+import munit._
+import org.scalacheck.Prop._
+import org.typelevel.idna4s.core.uts46._
+import com.ibm.icu.text.Normalizer2
+import java.lang.StringBuilder
+import org.typelevel.idna4s.core.uts46.CodePointMapper._
+
+trait CodePointMapperPlatformTests extends DisciplineSuite {
+  import CodePointMapperPlatformTests._
+
+  // Note this test will become flaky idna4's and icu4j are targeting
+  // different versions of Unicode.
+  property("idna4s's uts-46 mapping step, should agree with icu4j's uts46-mapping step"){
```

Comment:
Is it already flaky, or will it become flaky? I expect the icu4j version will have to be pinned.

Reply:
It may already be flaky. icu4j is using Unicode 14.0.0, [but as of September 13th 2022, we are using Unicode 15.0.0](https://www.unicode.org/versions/Unicode15.0.0/). 

icu4j has a `release-72-rc` branch that is on Unicode 15.0.0, but I don't see it on Maven central.

Reply:
Looks like there are changes in the mapping table in this revision. 

```diff
2,3c2,3
< # Date: 2021-07-10, 00:49:51 GMT
< # © 2021 Unicode®, Inc.
---
> # Date: 2022-05-02, 19:29:26 GMT
> # © 2022 Unicode®, Inc.
5c5
< # For terms of use, see http://www.unicode.org/terms_of_use.html
---
> # For terms of use, see https://www.unicode.org/terms_of_use.html
8c8
< # Version: 14.0.0
---
> # Version: 15.0.0
10c10
< # For documentation and usage, see http://www.unicode.org/reports/tr46
---
> # For documentation and usage, see https://www.unicode.org/reports/tr46
1281c1281,1282
< 0CF3..0CFF    ; disallowed                             # NA   <reserved-0CF3>..<reserved-0CFF>
---
> 0CF3          ; valid                                  # 15.0 KANNADA SIGN COMBINING ANUSVARA ABOVE RIGHT
> 0CF4..0CFF    ; disallowed                             # NA   <reserved-0CF4>..<reserved-0CFF>
1389c1390,1391
< 0ECE..0ECF    ; disallowed                             # NA   <reserved-0ECE>..<reserved-0ECF>
---
> 0ECE          ; valid                                  # 15.0 LAO YAMAKKAN
> 0ECF          ; disallowed                             # NA   <reserved-0ECF>
6209c6211,6212
< 10EB2..10EFF  ; disallowed                             # NA   <reserved-10EB2>..<reserved-10EFF>
---
> 10EB2..10EFC  ; disallowed                             # NA   <reserved-10EB2>..<reserved-10EFC>
> 10EFD..10EFF  ; valid                                  # 15.0 ARABIC SMALL LOW WORD SAKTA..ARABIC SMALL LOW WORD MADDA
6274c6277,6278
< 1123F..1127F  ; disallowed                             # NA   <reserved-1123F>..<reserved-1127F>
---
> 1123F..11241  ; valid                                  # 15.0 KHOJKI LETTER QA..KHOJKI VOWEL SIGN VOCALIC R
> 11242..1127F  ; disallowed                             # NA   <reserved-11242>..<reserved-1127F>
6446c6450,6452
< 11AF9..11BFF  ; disallowed                             # NA   <reserved-11AF9>..<reserved-11BFF>
---
> 11AF9..11AFF  ; disallowed                             # NA   <reserved-11AF9>..<reserved-11AFF>
> 11B00..11B09  ; valid                  ;      ; NV8    # 15.0 DEVANAGARI HEAD MARK..DEVANAGARI SIGN MINDU
> 11B0A..11BFF  ; disallowed                             # NA   <reserved-11B0A>..<reserved-11BFF>
6492c6498,6506
< 11EF9..11FAF  ; disallowed                             # NA   <reserved-11EF9>..<reserved-11FAF>
---
> 11EF9..11EFF  ; disallowed                             # NA   <reserved-11EF9>..<reserved-11EFF>
> 11F00..11F10  ; valid                                  # 15.0 KAWI SIGN CANDRABINDU..KAWI LETTER O
> 11F11         ; disallowed                             # NA   <reserved-11F11>
> 11F12..11F3A  ; valid                                  # 15.0 KAWI LETTER KA..KAWI VOWEL SIGN VOCALIC R
> 11F3B..11F3D  ; disallowed                             # NA   <reserved-11F3B>..<reserved-11F3D>
> 11F3E..11F42  ; valid                                  # 15.0 KAWI VOWEL SIGN E..KAWI CONJOINER
> 11F43..11F4F  ; valid                  ;      ; NV8    # 15.0 KAWI DANDA..KAWI PUNCTUATION CLOSING SPIRAL
> 11F50..11F59  ; valid                                  # 15.0 KAWI DIGIT ZERO..KAWI DIGIT NINE
> 11F5A..11FAF  ; disallowed                             # NA   <reserved-11F5A>..<reserved-11FAF>
6514c6528
< 1342F         ; disallowed                             # NA   <reserved-1342F>
---
> 1342F         ; valid                                  # 15.0 EGYPTIAN HIEROGLYPH V011D
6516c6530,6532
< 13439..143FF  ; disallowed                             # NA   <reserved-13439>..<reserved-143FF>
---
> 13439..1343F  ; disallowed                             # 15.0 EGYPTIAN HIEROGLYPH INSERT AT MIDDLE..EGYPTIAN HIEROGLYPH END WALLED ENCLOSURE
> 13440..13455  ; valid                                  # 15.0 EGYPTIAN HIEROGLYPH MIRROR HORIZONTALLY..EGYPTIAN HIEROGLYPH MODIFIER DAMAGED
> 13456..143FF  ; disallowed                             # NA   <reserved-13456>..<reserved-143FF>
6618c6634,6636
< 1B123..1B14F  ; disallowed                             # NA   <reserved-1B123>..<reserved-1B14F>
---
> 1B123..1B131  ; disallowed                             # NA   <reserved-1B123>..<reserved-1B131>
> 1B132         ; valid                                  # 15.0 HIRAGANA LETTER SMALL KO
> 1B133..1B14F  ; disallowed                             # NA   <reserved-1B133>..<reserved-1B14F>
6620c6638,6640
< 1B153..1B163  ; disallowed                             # NA   <reserved-1B153>..<reserved-1B163>
---
> 1B153..1B154  ; disallowed                             # NA   <reserved-1B153>..<reserved-1B154>
> 1B155         ; valid                                  # 15.0 KATAKANA LETTER SMALL KO
> 1B156..1B163  ; disallowed                             # NA   <reserved-1B156>..<reserved-1B163>
6671c6691,6693
< 1D246..1D2DF  ; disallowed                             # NA   <reserved-1D246>..<reserved-1D2DF>
---
> 1D246..1D2BF  ; disallowed                             # NA   <reserved-1D246>..<reserved-1D2BF>
> 1D2C0..1D2D3  ; valid                  ;      ; NV8    # 15.0 KAKTOVIK NUMERAL ZERO..KAKTOVIK NUMERAL NINETEEN
> 1D2D4..1D2DF  ; disallowed                             # NA   <reserved-1D2D4>..<reserved-1D2DF>
7704c7726,7728
< 1DF1F..1DFFF  ; disallowed                             # NA   <reserved-1DF1F>..<reserved-1DFFF>
---
> 1DF1F..1DF24  ; disallowed                             # NA   <reserved-1DF1F>..<reserved-1DF24>
> 1DF25..1DF2A  ; valid                                  # 15.0 LATIN SMALL LETTER D WITH MID-HEIGHT LEFT HOOK..LATIN SMALL LETTER T WITH MID-HEIGHT LEFT HOOK
> 1DF2B..1DFFF  ; disallowed                             # NA   <reserved-1DF2B>..<reserved-1DFFF>
7714c7738,7803
< 1E02B..1E0FF  ; disallowed                             # NA   <reserved-1E02B>..<reserved-1E0FF>
---
> 1E02B..1E02F  ; disallowed                             # NA   <reserved-1E02B>..<reserved-1E02F>
> 1E030         ; mapped                 ; 0430          # 15.0 MODIFIER LETTER CYRILLIC SMALL A
> 1E031         ; mapped                 ; 0431          # 15.0 MODIFIER LETTER CYRILLIC SMALL BE
> 1E032         ; mapped                 ; 0432          # 15.0 MODIFIER LETTER CYRILLIC SMALL VE
> 1E033         ; mapped                 ; 0433          # 15.0 MODIFIER LETTER CYRILLIC SMALL GHE
> 1E034         ; mapped                 ; 0434          # 15.0 MODIFIER LETTER CYRILLIC SMALL DE
> 1E035         ; mapped                 ; 0435          # 15.0 MODIFIER LETTER CYRILLIC SMALL IE
> 1E036         ; mapped                 ; 0436          # 15.0 MODIFIER LETTER CYRILLIC SMALL ZHE
> 1E037         ; mapped                 ; 0437          # 15.0 MODIFIER LETTER CYRILLIC SMALL ZE
> 1E038         ; mapped                 ; 0438          # 15.0 MODIFIER LETTER CYRILLIC SMALL I
> 1E039         ; mapped                 ; 043A          # 15.0 MODIFIER LETTER CYRILLIC SMALL KA
> 1E03A         ; mapped                 ; 043B          # 15.0 MODIFIER LETTER CYRILLIC SMALL EL
> 1E03B         ; mapped                 ; 043C          # 15.0 MODIFIER LETTER CYRILLIC SMALL EM
> 1E03C         ; mapped                 ; 043E          # 15.0 MODIFIER LETTER CYRILLIC SMALL O
> 1E03D         ; mapped                 ; 043F          # 15.0 MODIFIER LETTER CYRILLIC SMALL PE
> 1E03E         ; mapped                 ; 0440          # 15.0 MODIFIER LETTER CYRILLIC SMALL ER
> 1E03F         ; mapped                 ; 0441          # 15.0 MODIFIER LETTER CYRILLIC SMALL ES
> 1E040         ; mapped                 ; 0442          # 15.0 MODIFIER LETTER CYRILLIC SMALL TE
> 1E041         ; mapped                 ; 0443          # 15.0 MODIFIER LETTER CYRILLIC SMALL U
> 1E042         ; mapped                 ; 0444          # 15.0 MODIFIER LETTER CYRILLIC SMALL EF
> 1E043         ; mapped                 ; 0445          # 15.0 MODIFIER LETTER CYRILLIC SMALL HA
> 1E044         ; mapped                 ; 0446          # 15.0 MODIFIER LETTER CYRILLIC SMALL TSE
> 1E045         ; mapped                 ; 0447          # 15.0 MODIFIER LETTER CYRILLIC SMALL CHE
> 1E046         ; mapped                 ; 0448          # 15.0 MODIFIER LETTER CYRILLIC SMALL SHA
> 1E047         ; mapped                 ; 044B          # 15.0 MODIFIER LETTER CYRILLIC SMALL YERU
> 1E048         ; mapped                 ; 044D          # 15.0 MODIFIER LETTER CYRILLIC SMALL E
> 1E049         ; mapped                 ; 044E          # 15.0 MODIFIER LETTER CYRILLIC SMALL YU
> 1E04A         ; mapped                 ; A689          # 15.0 MODIFIER LETTER CYRILLIC SMALL DZZE
> 1E04B         ; mapped                 ; 04D9          # 15.0 MODIFIER LETTER CYRILLIC SMALL SCHWA
> 1E04C         ; mapped                 ; 0456          # 15.0 MODIFIER LETTER CYRILLIC SMALL BYELORUSSIAN-UKRAINIAN I
> 1E04D         ; mapped                 ; 0458          # 15.0 MODIFIER LETTER CYRILLIC SMALL JE
> 1E04E         ; mapped                 ; 04E9          # 15.0 MODIFIER LETTER CYRILLIC SMALL BARRED O
> 1E04F         ; mapped                 ; 04AF          # 15.0 MODIFIER LETTER CYRILLIC SMALL STRAIGHT U
> 1E050         ; mapped                 ; 04CF          # 15.0 MODIFIER LETTER CYRILLIC SMALL PALOCHKA
> 1E051         ; mapped                 ; 0430          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER A
> 1E052         ; mapped                 ; 0431          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER BE
> 1E053         ; mapped                 ; 0432          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER VE
> 1E054         ; mapped                 ; 0433          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER GHE
> 1E055         ; mapped                 ; 0434          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER DE
> 1E056         ; mapped                 ; 0435          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER IE
> 1E057         ; mapped                 ; 0436          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER ZHE
> 1E058         ; mapped                 ; 0437          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER ZE
> 1E059         ; mapped                 ; 0438          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER I
> 1E05A         ; mapped                 ; 043A          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER KA
> 1E05B         ; mapped                 ; 043B          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER EL
> 1E05C         ; mapped                 ; 043E          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER O
> 1E05D         ; mapped                 ; 043F          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER PE
> 1E05E         ; mapped                 ; 0441          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER ES
> 1E05F         ; mapped                 ; 0443          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER U
> 1E060         ; mapped                 ; 0444          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER EF
> 1E061         ; mapped                 ; 0445          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER HA
> 1E062         ; mapped                 ; 0446          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER TSE
> 1E063         ; mapped                 ; 0447          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER CHE
> 1E064         ; mapped                 ; 0448          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER SHA
> 1E065         ; mapped                 ; 044A          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER HARD SIGN
> 1E066         ; mapped                 ; 044B          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER YERU
> 1E067         ; mapped                 ; 0491          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER GHE WITH UPTURN
> 1E068         ; mapped                 ; 0456          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER BYELORUSSIAN-UKRAINIAN I
> 1E069         ; mapped                 ; 0455          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER DZE
> 1E06A         ; mapped                 ; 045F          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER DZHE
> 1E06B         ; mapped                 ; 04AB          # 15.0 MODIFIER LETTER CYRILLIC SMALL ES WITH DESCENDER
> 1E06C         ; mapped                 ; A651          # 15.0 MODIFIER LETTER CYRILLIC SMALL YERU WITH BACK YER
> 1E06D         ; mapped                 ; 04B1          # 15.0 MODIFIER LETTER CYRILLIC SMALL STRAIGHT U WITH STROKE
> 1E06E..1E08E  ; disallowed                             # NA   <reserved-1E06E>..<reserved-1E08E>
> 1E08F         ; valid                                  # 15.0 COMBINING CYRILLIC SMALL LETTER BYELORUSSIAN-UKRAINIAN I
> 1E090..1E0FF  ; disallowed                             # NA   <reserved-1E090>..<reserved-1E0FF>
7729c7818,7820
< 1E300..1E7DF  ; disallowed                             # NA   <reserved-1E300>..<reserved-1E7DF>
---
> 1E300..1E4CF  ; disallowed                             # NA   <reserved-1E300>..<reserved-1E4CF>
> 1E4D0..1E4F9  ; valid                                  # 15.0 NAG MUNDARI LETTER O..NAG MUNDARI DIGIT NINE
> 1E4FA..1E7DF  ; disallowed                             # NA   <reserved-1E4FA>..<reserved-1E7DF>
8216c8307,8308
< 1F6D8..1F6DC  ; disallowed                             # NA   <reserved-1F6D8>..<reserved-1F6DC>
---
> 1F6D8..1F6DB  ; disallowed                             # NA   <reserved-1F6D8>..<reserved-1F6DB>
> 1F6DC         ; valid                  ;      ; NV8    # 15.0 WIRELESS
8228c8320,8322
< 1F774..1F77F  ; disallowed                             # NA   <reserved-1F774>..<reserved-1F77F>
---
> 1F774..1F776  ; valid                  ;      ; NV8    # 15.0 LOT OF FORTUNE..LUNAR ECLIPSE
> 1F777..1F77A  ; disallowed                             # NA   <reserved-1F777>..<reserved-1F77A>
> 1F77B..1F77F  ; valid                  ;      ; NV8    # 15.0 HAUMEA..ORCUS
8231c8325,8326
< 1F7D9..1F7DF  ; disallowed                             # NA   <reserved-1F7D9>..<reserved-1F7DF>
---
> 1F7D9         ; valid                  ;      ; NV8    # 15.0 NINE POINTED WHITE STAR
> 1F7DA..1F7DF  ; disallowed                             # NA   <reserved-1F7DA>..<reserved-1F7DF>
8298c8393
< 1FA75..1FA77  ; disallowed                             # NA   <reserved-1FA75>..<reserved-1FA77>
---
> 1FA75..1FA77  ; valid                  ;      ; NV8    # 15.0 LIGHT BLUE HEART..PINK HEART
8304c8399,8400
< 1FA87..1FA8F  ; disallowed                             # NA   <reserved-1FA87>..<reserved-1FA8F>
---
> 1FA87..1FA88  ; valid                  ;      ; NV8    # 15.0 MARACAS..FLUTE
> 1FA89..1FA8F  ; disallowed                             # NA   <reserved-1FA89>..<reserved-1FA8F>
8308c8404
< 1FAAD..1FAAF  ; disallowed                             # NA   <reserved-1FAAD>..<reserved-1FAAF>
---
> 1FAAD..1FAAF  ; valid                  ;      ; NV8    # 15.0 FOLDING HAND FAN..KHANDA
8311c8407,8409
< 1FABB..1FABF  ; disallowed                             # NA   <reserved-1FABB>..<reserved-1FABF>
---
> 1FABB..1FABD  ; valid                  ;      ; NV8    # 15.0 HYACINTH..WING
> 1FABE         ; disallowed                             # NA   <reserved-1FABE>
> 1FABF         ; valid                  ;      ; NV8    # 15.0 GOOSE
8314c8412,8413
< 1FAC6..1FACF  ; disallowed                             # NA   <reserved-1FAC6>..<reserved-1FACF>
---
> 1FAC6..1FACD  ; disallowed                             # NA   <reserved-1FAC6>..<reserved-1FACD>
> 1FACE..1FACF  ; valid                  ;      ; NV8    # 15.0 MOOSE..DONKEY
8317c8416,8417
< 1FADA..1FADF  ; disallowed                             # NA   <reserved-1FADA>..<reserved-1FADF>
---
> 1FADA..1FADB  ; valid                  ;      ; NV8    # 15.0 GINGER ROOT..PEA POD
> 1FADC..1FADF  ; disallowed                             # NA   <reserved-1FADC>..<reserved-1FADF>
8319c8419,8420
< 1FAE8..1FAEF  ; disallowed                             # NA   <reserved-1FAE8>..<reserved-1FAEF>
---
> 1FAE8         ; valid                  ;      ; NV8    # 15.0 SHAKING FACE
> 1FAE9..1FAEF  ; disallowed                             # NA   <reserved-1FAE9>..<reserved-1FAEF>
8321c8422,8423
< 1FAF7..1FAFF  ; disallowed                             # NA   <reserved-1FAF7>..<reserved-1FAFF>
---
> 1FAF7..1FAF8  ; valid                  ;      ; NV8    # 15.0 LEFTWARDS PUSHING HAND..RIGHTWARDS PUSHING HAND
> 1FAF9..1FAFF  ; disallowed                             # NA   <reserved-1FAF9>..<reserved-1FAFF>
8344c8446,8447
< 2B739..2B73F  ; disallowed                             # NA   <reserved-2B739>..<reserved-2B73F>
---
> 2B739         ; valid                                  # 15.0 CJK UNIFIED IDEOGRAPH-2B739
> 2B73A..2B73F  ; disallowed                             # NA   <reserved-2B73A>..<reserved-2B73F>
8886c8989,8991
< 3134B..3FFFD  ; disallowed                             # NA   <reserved-3134B>..<reserved-3FFFD>
---
> 3134B..3134F  ; disallowed                             # NA   <reserved-3134B>..<reserved-3134F>
> 31350..323AF  ; valid                                  # 15.0 CJK UNIFIED IDEOGRAPH-31350..CJK UNIFIED IDEOGRAPH-323AF
> 323B0..3FFFD  ; disallowed                             # NA   <reserved-323B0>..<reserved-3FFFD>
```

Reply:
That is, some code points which our version of UTS-46 consider valid (because they are under Unicode 15.0.0), icu4j 71.0 considers disallowed (because they are under Unicode 14.0.0)

Reply:
Seems they have a tagged release as well:
https://github.com/unicode-org/icu/releases/tag/release-72-rc

Hopefully they'll have some published artifacts for that soon.